S3TransferUtilitySample/src/com/amazonaws/demo/s3transferutility/UploadActivity.java

```diff
@@ -519,85 +517,68 @@ private void beginUploadInBackground(String filePath) {
     }
 
     /*
-     * Gets the file path of the given Uri.
+     * Copies the resource associated with the Uri to a new File in the cache directory.
+     * @param uri The Uri.
+     * @return a new File representing the Uri in the cache directory.
+
      */
-    @SuppressLint("NewApi")
-    private String getPath(Uri uri) throws URISyntaxException {
-        final boolean needToCheckUri = Build.VERSION.SDK_INT >= 19;
-        String selection = null;
-        String[] selectionArgs = null;
-        // Uri is different in versions after KITKAT (Android 4.4), we need to
-        // deal with different Uris.
-        if (needToCheckUri && DocumentsContract.isDocumentUri(getApplicationContext(), uri)) {
-            if (isExternalStorageDocument(uri)) {
-                final String docId = DocumentsContract.getDocumentId(uri);
-                final String[] split = docId.split(":");
-                return Environment.getExternalStorageDirectory() + "/" + split[1];
-            } else if (isDownloadsDocument(uri)) {
-                final String id = DocumentsContract.getDocumentId(uri);
-                uri = ContentUris.withAppendedId(
-                        Uri.parse("content://downloads/public_downloads"), Long.valueOf(id));
-            } else if (isMediaDocument(uri)) {
-                final String docId = DocumentsContract.getDocumentId(uri);
-                final String[] split = docId.split(":");
-                final String type = split[0];
-                if ("image".equals(type)) {
-                    uri = MediaStore.Images.Media.EXTERNAL_CONTENT_URI;
-                } else if ("video".equals(type)) {
-                    uri = MediaStore.Video.Media.EXTERNAL_CONTENT_URI;
-                } else if ("audio".equals(type)) {
-                    uri = MediaStore.Audio.Media.EXTERNAL_CONTENT_URI;
+    private File getFile(Uri uri) throws FileNotFoundException {
+        final InputStream in = getContentResolver().openInputStream(uri);
+        OutputStream out = null;
+        try {
```

Comment:
You can use the "try with resources",  and scope `in` and `out` to the `try` block in such a way that they auto-close.

https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html

^ @raphkim and I recently discovered that this now gets desugared by Android -- I think for all versions? (It didn't used to be this, on Andorid, try-with-resources was sort of just an Oracle Java thing.)

Another advantage of that, here, is that you won't have to worry about these streams being set to `null`, and could make the assignments as `final`.

Reply:
TIL :). Thanks for this, this is much cleaner.